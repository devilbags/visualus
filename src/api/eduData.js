import axios from 'axios';

export const apiEndpoint =
  'https://api.data.gov/ed/collegescorecard/v1/schools.json';

export const universityList = input =>
  new Promise((resolve, reject) => {
    axios
      .get(apiEndpoint, {
        params: Object.assign(
          {},
          {
            api_key: process.env.REACT_APP_GOV_DATA_API_KEY, // Appending api key
            'school.name': input, // Append user input
            _fields:
              'id,school.name,school.alias,school.city,school.state,school.school_url,2015.student.size,2015.admissions.admission_rate.overall,2015.cost.attendance.academic_year'
          }
        )
      })
      .then(response => {
        const list = response.data.results.map(school => {
          const id = school['id'];
          const name = school['school.name'];
          const alias = school['school.alias'];
          const city = school['school.city'];
          const state = school['school.state'];
          const website = school['school.school_url'];
          const studentSize = school['2015.student.size'];
          const admissionRate =
            school['2015.admissions.admission_rate.overall'];
          const annualCost = school['2015.cost.attendance.academic_year'];
          return {
            id: id,
            name: name,
            alias: alias,
            city: city,
            state: state,
            website: website,
            studentSize: studentSize,
            admissionRate: admissionRate,
            annualCost: annualCost
          };
        });
        resolve(list);
      })
      .catch(error => {
        reject(error);
      });
  });
