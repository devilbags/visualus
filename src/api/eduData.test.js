import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { apiEndpoint, universityList, universityDetail } from './eduData';

describe('Government Education API', () => {
  it('can fetch', async () => {
    expect.assertions(1);
    const mock = new MockAdapter(axios);
    mock.onGet(apiEndpoint).reply(200, {
      metadata: {
        total: 1,
        page: 0,
        per_page: 1
      },
      results: [
        {
          'school.name': 'University of California-Los Angeles',
          'school.alias': 'UCLA',
          id: 110662
        }
      ]
    });
    const response = await universityList('uc');
    expect(response[0].id).toEqual(110662);
  });
  it('can handle error', async () => {
    const mock = new MockAdapter(axios);
    mock.onGet(apiEndpoint).networkError();
    await universityList().catch(error => {
      expect(error).toMatchSnapshot();
    });
  });
  it('can fetch single university detail', async () => {
    expect.assertions(2);
    const mock = new MockAdapter(axios);
    mock.onGet(apiEndpoint).reply(200, {
      metadata: {},
      results: [
        {
          '2013.student.demographics.race_ethnicity.non_resident_alien': 0.1051,
          '2010.student.demographics.men': 0.4962,
          '2009.aid.students_with_any_loan': 0.5790816327,
          '2010.student.demographics.women': 0.5038,
          '2010.student.demographics.race_ethnicity.asian': 0.1414,
          '2007.cost.tuition.out_of_state': 34530,
          '2012.admissions.admission_rate.overall': 0.0773,
          '2009.student.demographics.race_ethnicity.asian': 0,
          '2015.aid.students_with_any_loan': 0.4772727273,
          '2014.student.demographics.race_ethnicity.asian': 0.1655,
          '2012.student.demographics.race_ethnicity.asian': 0.1608,
          '2010.student.demographics.race_ethnicity.non_resident_alien': 0.0982,
          '2010.student.demographics.race_ethnicity.black': 0.0597,
          '2009.student.grad_students': 6318,
          '2012.student.size': 5397,
          '2015.admissions.sat_scores.average.overall': 1498,
          '2008.student.size': 5256,
          '2014.student.demographics.men': 0.5098,
          '2008.completion.completion_rate_4yr_100nt': 0.8960739030023,
          '2007.student.demographics.race_ethnicity.black_non_hispanic': 0.0877,
          id: 130794,
          '2010.completion.completion_rate_4yr_100nt': 0.88897396630934,
          '2008.admissions.sat_scores.average.overall': 1481,
          '2009.student.demographics.men': 0.5013,
          '2010.cost.tuition.in_state': 38300,
          '2007.student.demographics.race_ethnicity.white_non_hispanic': 0.4869,
          '2013.aid.students_with_any_loan': 0.561827957,
          '2005.cost.tuition.out_of_state': 31460,
          '2008.student.grad_students': 4915,
          '2006.student.demographics.women': 0.4919,
          '2008.student.demographics.race_ethnicity.non_resident_alien': 0.086,
          '2009.student.demographics.race_ethnicity.white': 0,
          '2013.aid.loan_principal': 11000,
          '2011.cost.tuition.in_state': 40500,
          '2012.admissions.sat_scores.average.overall': 1488,
          '2013.admissions.admission_rate.overall': 0.0705,
          '2012.cost.tuition.in_state': 42300,
          '2014.admissions.admission_rate.overall': 0.063,
          '2012.student.demographics.race_ethnicity.black': 0.0656,
          '2014.student.demographics.women': 0.4902,
          '2014.student.demographics.race_ethnicity.black': 0.0678,
          '2007.student.grad_students': 4895,
          '2007.aid.loan_principal': 11625,
          '2006.aid.loan_principal': 11721,
          '2007.aid.students_with_any_loan': 0.8813793103,
          '2010.aid.students_with_any_loan': 0.6253132832,
          '2009.cost.tuition.in_state': 36500,
          '2014.cost.tuition.out_of_state': 45800,
          '2015.student.demographics.women': 0.4884,
          '2013.cost.tuition.in_state': 44000,
          '2009.student.demographics.women': 0.4987,
          '2005.student.demographics.race_ethnicity.black_non_hispanic': 0.0811,
          '2015.student.size': 5528,
          '2013.student.size': 5422,
          '2005.student.size': 5349,
          '2005.completion.completion_rate_4yr_100nt': 0.89764359351988,
          '2015.cost.tuition.out_of_state': 47600,
          '2005.student.grad_students': 4828,
          '2008.admissions.admission_rate.overall': 0.0989,
          '2010.admissions.admission_rate.overall': 0.0753,
          '2007.admissions.sat_scores.average.overall': 1471,
          '2007.student.size': 5291,
          '2011.student.size': 5333,
          '2009.student.demographics.race_ethnicity.white_non_hispanic': 0.4167,
          '2014.student.demographics.race_ethnicity.non_resident_alien': 0.104,
          '2015.admissions.admission_rate.overall': 0.0673,
          '2009.student.size': 5258,
          '2007.cost.tuition.in_state': 34530,
          '2015.cost.tuition.in_state': 47600,
          '2008.student.avg_dependent_income.2014dollars': 87601,
          '2009.cost.tuition.out_of_state': 36500,
          '2010.student.avg_dependent_income.2014dollars': 94031,
          '2013.admissions.sat_scores.average.overall': 1497,
          '2011.student.demographics.race_ethnicity.white': 0.4723,
          '2012.student.demographics.race_ethnicity.white': 0.4719,
          '2014.student.demographics.race_ethnicity.white': 0.4705,
          '2015.student.demographics.race_ethnicity.white': 0.4651,
          '2015.student.demographics.race_ethnicity.non_resident_alien': 0.106,
          '2011.student.grad_students': 6526,
          '2007.completion.completion_rate_4yr_100nt': 0.87006960556844,
          '2006.student.avg_dependent_income.2014dollars': 94813,
          '2013.student.demographics.race_ethnicity.white': 0.4734,
          '2012.student.avg_dependent_income.2014dollars': 92921,
          '2014.student.avg_dependent_income.2014dollars': 95354,
          '2014.aid.students_with_any_loan': 0.5576662144,
          '2010.student.demographics.race_ethnicity.white': 0.4666,
          '2010.cost.tuition.out_of_state': 38300,
          '2014.student.grad_students': 6859,
          '2006.completion.completion_rate_4yr_100nt': 0.88139362490733,
          '2005.cost.tuition.in_state': 31460,
          '2005.aid.loan_principal': 12766.5,
          '2015.student.demographics.race_ethnicity.asian': 0.1729,
          '2007.student.demographics.men': 0.5101,
          '2008.student.demographics.race_ethnicity.asian': 0,
          '2014.student.size': 5473,
          '2011.student.demographics.race_ethnicity.black': 0.063,
          '2011.student.demographics.race_ethnicity.asian': 0.1523,
          '2013.student.demographics.race_ethnicity.asian': 0.1645,
          '2005.student.demographics.women': 0.4907,
          '2008.cost.tuition.out_of_state': 35300,
          '2011.aid.loan_principal': 7822.5,
          '2015.completion.completion_rate_4yr_100nt': 0.8774,
          '2006.student.size': 5302,
          '2010.student.size': 5296,
          '2008.aid.students_with_any_loan': 0.7426273458,
          '2006.admissions.sat_scores.average.overall': 1474,
          '2012.student.demographics.men': 0.5027,
          '2013.student.grad_students': 6679,
          '2006.admissions.admission_rate.overall': 0.0967,
          '2006.student.demographics.race_ethnicity.black_non_hispanic': 0.0824,
          '2009.completion.completion_rate_4yr_100nt': 0.90444444444444,
          '2015.aid.loan_principal': 13000,
          '2008.student.demographics.race_ethnicity.white': 0,
          '2014.aid.loan_principal': 13000,
          '2014.admissions.sat_scores.average.overall': 1493,
          '2006.cost.tuition.in_state': 33030,
          '2015.student.grad_students': 6853,
          '2006.cost.tuition.out_of_state': 33030,
          '2013.student.demographics.race_ethnicity.black': 0.0673,
          '2010.admissions.sat_scores.average.overall': 1473,
          '2005.admissions.admission_rate.overall': 0.0995,
          '2011.student.demographics.women': 0.5014,
          '2008.student.demographics.race_ethnicity.white_non_hispanic': 0.4509,
          '2009.student.demographics.race_ethnicity.non_resident_alien': 0.0888,
          '2012.student.demographics.race_ethnicity.non_resident_alien': 0.1026,
          '2015.student.demographics.race_ethnicity.black': 0.068,
          '2008.aid.loan_principal': 10704,
          '2009.aid.loan_principal': 8380,
          '2011.completion.completion_rate_4yr_100nt': 0.88787878787878,
          '2013.student.avg_dependent_income.2014dollars': 91847,
          '2011.aid.students_with_any_loan': 0.6393861893,
          '2008.student.demographics.men': 0.5013,
          '2006.student.grad_students': 4852,
          '2008.student.demographics.women': 0.4987,
          '2010.student.grad_students': 6391,
          '2009.admissions.admission_rate.overall': 0.0856,
          '2010.aid.loan_principal': 7125,
          '2011.student.demographics.men': 0.4986,
          '2007.student.demographics.race_ethnicity.non_resident_alien': 0.0839,
          '2013.cost.tuition.out_of_state': 44000,
          '2013.student.demographics.men': 0.5063,
          '2012.aid.students_with_any_loan': 0.5778688525,
          '2014.cost.tuition.in_state': 45800,
          '2008.student.demographics.race_ethnicity.black_non_hispanic': 0.0869,
          '2012.aid.loan_principal': 9858,
          '2008.cost.tuition.in_state': 35300,
          '2005.admissions.sat_scores.average.overall': 1471,
          '2011.student.demographics.race_ethnicity.non_resident_alien': 0.1014,
          '2006.student.demographics.race_ethnicity.white_non_hispanic': 0.5004,
          '2014.completion.completion_rate_4yr_100nt': 0.8739,
          '2005.aid.students_with_any_loan': 0.9439953811,
          '2009.admissions.sat_scores.average.overall': 1475,
          '2012.cost.tuition.out_of_state': 42300,
          '2012.student.grad_students': 6501,
          '2005.student.demographics.race_ethnicity.non_resident_alien': 0.0817,
          '2013.student.demographics.women': 0.4937,
          '2009.student.avg_dependent_income.2014dollars': 87518,
          '2007.student.avg_dependent_income.2014dollars': 93040,
          '2013.completion.completion_rate_4yr_100nt': 0.8964,
          '2012.student.demographics.women': 0.4973,
          '2006.aid.students_with_any_loan': 0.9344675488,
          '2015.student.demographics.men': 0.5116,
          '2006.student.demographics.men': 0.5081,
          '2005.student.demographics.race_ethnicity.white_non_hispanic': 0.5106,
          '2011.cost.tuition.out_of_state': 40500,
          '2006.student.demographics.race_ethnicity.non_resident_alien': 0.083,
          '2011.admissions.sat_scores.average.overall': 1497,
          '2007.student.demographics.women': 0.4899,
          '2005.student.avg_dependent_income.2014dollars': 89241,
          '2011.student.avg_dependent_income.2014dollars': 96192,
          '2015.student.avg_dependent_income.2014dollars': 89141,
          '2008.student.demographics.race_ethnicity.black': 0,
          '2007.admissions.admission_rate.overall': 0.089,
          '2009.student.demographics.race_ethnicity.black': 0,
          '2009.student.demographics.race_ethnicity.black_non_hispanic': 0.0852,
          '2012.completion.completion_rate_4yr_100nt': 0.89032749428789,
          '2005.student.demographics.men': 0.5093,
          '2011.admissions.admission_rate.overall': 0.0788,
          '2005.student.demographics.race_ethnicity.black': null,
          '2006.student.demographics.race_ethnicity.black': null,
          '2007.student.demographics.race_ethnicity.black': null,
          '2010.student.demographics.race_ethnicity.black_non_hispanic': null,
          '2011.student.demographics.race_ethnicity.black_non_hispanic': null,
          '2012.student.demographics.race_ethnicity.black_non_hispanic': null,
          '2013.student.demographics.race_ethnicity.black_non_hispanic': null,
          '2014.student.demographics.race_ethnicity.black_non_hispanic': null,
          '2015.student.demographics.race_ethnicity.black_non_hispanic': null,
          '2005.student.demographics.race_ethnicity.asian': null,
          '2006.student.demographics.race_ethnicity.asian': null,
          '2007.student.demographics.race_ethnicity.asian': null,
          '2005.student.demographics.race_ethnicity.white': null,
          '2006.student.demographics.race_ethnicity.white': null,
          '2007.student.demographics.race_ethnicity.white': null,
          '2010.student.demographics.race_ethnicity.white_non_hispanic': null,
          '2011.student.demographics.race_ethnicity.white_non_hispanic': null,
          '2012.student.demographics.race_ethnicity.white_non_hispanic': null,
          '2013.student.demographics.race_ethnicity.white_non_hispanic': null,
          '2014.student.demographics.race_ethnicity.white_non_hispanic': null,
          '2015.student.demographics.race_ethnicity.white_non_hispanic': null
        }
      ]
    });
    const resp = await universityDetail(1);
    expect(resp.id).toEqual(1);
    expect(resp.detail.cost.inState['2005']).toEqual(31460);
  });
  it('fetching detail can handle error', async () => {
    const mock = new MockAdapter(axios);
    mock.onGet(apiEndpoint).networkError();
    await universityDetail(1).catch(error => {
      expect(error).toMatchSnapshot();
    });
  });
});
