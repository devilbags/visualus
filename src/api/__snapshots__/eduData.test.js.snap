// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Goverment Education API can handle error 1`] = `[Error: Network Error]`;
